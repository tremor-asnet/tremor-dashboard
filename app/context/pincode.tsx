"use client";

import {
  ReactNode,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

interface IPinCodeContext {
  isConfirmPinCode: boolean;
  pinCode?: number;
  isShowPinCodeModal: boolean;
  setPinCode: (code: number) => void;
  showPinCodeModal: () => void;
  hidePinCodeModal: () => void;
  confirmPinCode: (code: number) => boolean;
}

const initialPinCodeContext: IPinCodeContext = {
  isConfirmPinCode: false,
  confirmPinCode: () => false,
  hidePinCodeModal: () => {},
  setPinCode: () => false,
  showPinCodeModal: () => {},
  isShowPinCodeModal: false,
};

const PinCodeContext = createContext(initialPinCodeContext);

const PinCodeProvider = ({ children }: { children: ReactNode }) => {
  const [isConfirmPinCode, setIsConfirmPinCode] = useState(false);
  const [isShowPinCodeModal, setIsShowPinCodeModal] = useState(false);
  const [pinCode, setPinCode] = useState<number>();

  useEffect(() => {
    setIsShowPinCodeModal(!pinCode);
  }, [pinCode]);

  const handleShowPinCodeModal = useCallback(
    () => setIsShowPinCodeModal(true),
    [],
  );

  const handleHidePinCodeModal = useCallback(
    () => setIsShowPinCodeModal(false),
    [],
  );

  const handleConfirmPinCode = useCallback(
    (code: number) => {
      const isMatch = code === pinCode;
      setIsConfirmPinCode(isMatch);
      isMatch ? handleHidePinCodeModal() : handleShowPinCodeModal();
      return isMatch;
    },
    [handleHidePinCodeModal, handleShowPinCodeModal, pinCode],
  );

  const pinCodeContextValue: IPinCodeContext = {
    isConfirmPinCode,
    confirmPinCode: handleConfirmPinCode,
    pinCode,
    isShowPinCodeModal,
    setPinCode,
    hidePinCodeModal: handleHidePinCodeModal,
    showPinCodeModal: handleShowPinCodeModal,
  };

  return (
    <PinCodeContext.Provider value={pinCodeContextValue}>
      {children}
    </PinCodeContext.Provider>
  );
};

const usePinCode = () => {
  const context = useContext(PinCodeContext);

  if (!context) {
    throw new Error("usePinCode hooks should using inside PinCodeProvider!");
  }

  return context;
};

export { PinCodeProvider, usePinCode };
