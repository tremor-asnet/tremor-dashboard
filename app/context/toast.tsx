"use client";

import React, { createContext, useState } from "react";

// Components
import { FaCheckCircle } from "react-icons/fa";
import { TbExclamationMark } from "react-icons/tb";
import { RxCross2 } from "react-icons/rx";
import Toast from "@/ui/components/Toast/Toast";

export enum TOAST_TYPE {
  SUCCESS = "success",
  WARNING = "warning",
  ERROR = "error",
}
interface ToastProps {
  type?: TOAST_TYPE;
  message?: string;
  icon?: JSX.Element;
}

type ToastState = ToastProps & { isOpen: boolean };

interface ToastContextValue {
  openToast: (toastProps: ToastProps) => void;
  closeToast: () => void;
}

export const buildToastRenderer = ({ type, icon, message }: ToastProps) => {
  switch (type) {
    case TOAST_TYPE.WARNING:
      return {
        icon: icon || TbExclamationMark,
        message: message ?? "Warning!",
        color: "yellow",
      };

    case TOAST_TYPE.ERROR:
      return {
        icon: icon || RxCross2,
        message: message ?? "Error!",
        color: "red",
      };

    case TOAST_TYPE.SUCCESS:
    default:
      return {
        icon: icon || FaCheckCircle,
        message: message ?? "Success!",
        color: "green",
      };
  }
};

const ToastContext = createContext<ToastContextValue>({
  openToast: () => {},
  closeToast: () => {},
});

const ToastProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [toast, setToast] = useState<ToastState>({
    isOpen: false,
    type: TOAST_TYPE.SUCCESS,
  });

  const closeToast = () => {
    setToast({
      ...toast,
      isOpen: false,
    });
  };

  const openToast = ({
    type = TOAST_TYPE.SUCCESS,
    message,
    icon,
  }: ToastProps) => {
    setToast({ isOpen: true, type, message, icon });

    setTimeout(() => {
      closeToast();
    }, 3000);
  };

  const toastRenderer = buildToastRenderer(toast);
  return (
    <ToastContext.Provider
      value={{
        openToast,
        closeToast,
      }}>
      {children}
      {toast.isOpen && (
        <Toast
          Icon={toastRenderer.icon}
          message={toastRenderer.message}
          color={toastRenderer.color}
          onClose={closeToast}
        />
      )}
    </ToastContext.Provider>
  );
};

export { ToastProvider, ToastContext };
