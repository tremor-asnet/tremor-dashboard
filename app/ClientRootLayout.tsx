"use client";

import { useContext } from "react";
import type { Metadata } from "next";
import { Inter } from "next/font/google";

// Styles
import "./globals.css";

// Contexts
import { ThemeContext } from "@/context/theme";

interface ClientRootLayoutProps {
  children: React.ReactNode;
}

export const metadata: Metadata = {
  title: "Tremor Dashboard",
  description: "Tremor Dashboard generated by create next app",
};

const inter = Inter({ subsets: ["latin"] });

const ClientRootLayout = ({ children }: ClientRootLayoutProps) => {
  const { isDarkTheme } = useContext(ThemeContext);

  return (
    <html lang="en">
      <head>
        <link rel="icon" href="/assets/images/favicon.ico" />
      </head>
      {/* TODO: Add class to custom theme dark or light */}
      <body className={`${inter.className} ${isDarkTheme ? "dark" : "light"}`}>
        {children}
        <div id="modal-root"></div>
      </body>
    </html>
  );
};

export default ClientRootLayout;
