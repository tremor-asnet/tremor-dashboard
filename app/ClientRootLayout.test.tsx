import { fireEvent, render } from "@testing-library/react";
import ClientRootLayout, { metadata } from "./ClientRootLayout";
import { ThemeContext, ThemeProvider } from "./context/theme";

describe("ClientRootLayout", () => {
  it("Should have correct title", () => {
    expect(metadata.title).toBe("Tremor Dashboard");
  });

  it("Should have correct description", () => {
    expect(metadata.description).toBe(
      "Tremor Dashboard generated by create next app",
    );
  });

  it("Should be correct with theme", () => {
    const { getByTestId, container } = render(
      <ThemeProvider>
        <ThemeContext.Consumer>
          {({ isDarkTheme, toggleTheme }) => (
            <ClientRootLayout>
              <button
                data-testid="toggle-theme"
                onClick={() =>
                  toggleTheme()
                }>{`isDarkTheme to ${isDarkTheme}`}</button>
            </ClientRootLayout>
          )}
        </ThemeContext.Consumer>
      </ThemeProvider>,
    );
    const themeButton = getByTestId("toggle-theme");

    fireEvent.click(themeButton);
    expect(themeButton.textContent).toBe("isDarkTheme to true");

    fireEvent.click(themeButton);
    expect(themeButton.textContent).toBe("isDarkTheme to false");

    expect(container).toMatchSnapshot();
  });
});
