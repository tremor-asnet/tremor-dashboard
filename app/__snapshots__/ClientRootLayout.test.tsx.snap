// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientRootLayout Should be correct with theme 1`] = `
<div>
  <html
    lang="en"
  >
    <head>
      <link
        href="/assets/images/favicon.ico"
        rel="icon"
      />
    </head>
    <body
      class="className light"
    >
      <button
        data-testid="toggle-theme"
      >
        isDarkTheme to false
      </button>
    </body>
  </html>
</div>
`;
